src/bench.cpp

```diff
@@ -0,0 +1,325 @@
+        cairo_stroke(cr);
+
+        cairo_set_source_rgba_swizzle(cr, 0.7, 0.7, 0.7, 0.7);
+        cairo_move_to(cr, xc, yc);
+        cairo_arc_negative(cr, xc, yc, radius, min_angle, max_angle);
+        cairo_fill(cr);
+
+        cairo_set_source_rgba_swizzle(cr, 1.0, 0.2, 0.2, 0.7);
+        cairo_move_to(cr, xc, yc);
+        cairo_arc_negative(cr, xc, yc, radius, fps_angle, max_angle);
+        cairo_fill(cr);
+
+        if (output->handle->current_mode)
+            cairo_set_source_rgba_swizzle(cr, 0, 0, 1, 1);
+        else
+            cairo_set_source_rgba_swizzle(cr, 1, 1, 0, 1);
+
+        cairo_text_extents (cr, fps_buf, &text_extents);
+        cairo_move_to(cr,
+            xc - text_extents.width / 2 + text_extents.x_bearing,
+            text_y + yc);
+        cairo_show_text(cr, fps_buf);
+        cairo_stroke(cr);
+    }
+
+    void update_fps()
+    {
+        uint32_t elapsed;
+        uint32_t current_time = wf::get_current_time();
+        elapsed = current_time - last_time;
+
+        last_time = current_time;
+
+        current_fps = (double) 1000 / elapsed;
+
+        if (current_fps > max_fps)
+            max_fps = current_fps;
+        else
+            max_fps -= 1;
+
+        sprintf(fps_buf, "%.1f", current_fps);
+    }
+
+    wf::effect_hook_t pre_hook = [=] ()
+    {
+        output->render->damage({
+            cairo_geometry.x, cairo_geometry.y,
+            cairo_geometry.width, cairo_geometry.height});
+    };
+
+    wf::effect_hook_t overlay_hook = [=] ()
+    {
+        update_fps();
+        if (frames++ >= frames_per_update)
+        {
+             render_bench();
+             frames = 0;
+        }
+
+        OpenGL::render_begin();
+        bench_tex.allocate(cairo_geometry.width, cairo_geometry.height);
+        GL_CALL(glBindTexture(GL_TEXTURE_2D, bench_tex.tex));
+        GL_CALL(glTexImage2D(GL_TEXTURE_2D, 0, GL_RGBA,
+            cairo_geometry.width, cairo_geometry.height,
+            0, GL_RGBA, GL_UNSIGNED_BYTE,
+            cairo_image_surface_get_data(cairo_surface)));
+        gl_geometry src_geometry = {
+            (float) cairo_geometry.x,
+            (float) cairo_geometry.y,
+            (float) cairo_geometry.x + cairo_geometry.width,
+            (float) cairo_geometry.y + cairo_geometry.height};
+        auto fb = output->render->get_target_framebuffer();
+        fb.bind();
+        OpenGL::render_transformed_texture(bench_tex.tex, src_geometry, {},
+            fb.get_orthographic_projection(), glm::vec4(1.0),
+            TEXTURE_TRANSFORM_INVERT_Y);
+        OpenGL::render_end();
+    };
+
+    void unset_hook()
+    {
+        if (!hook_set)
+            return;
+
+        output->render->set_redraw_always(false);
+        output->render->rem_effect(&pre_hook);
+        output->render->rem_effect(&overlay_hook);
+        hook_set = false;
+    }
+
+    void fini() override
+    {
+        unset_hook();
+        cairo_surface_destroy(cairo_surface);
```

Comment:
Can't it happen, in theory, that cairo-surface is nullptr? The only thing which is really guaranteed is that init() will be called before fini(), it may be that no frames are ever rendered (i.e on a noop output)

Reply:
It is not possible AFAICT. The call chain is: init->update_texture_position->cairo_recreate->cairo_image_surface_create. Thus, cairo-surface is set right up front.

Reply:
Ah yes, I missed this call in init.